### ISSUE TITLE
Clearly distinguish and document the two update mechanism in MFS: single-file upwards (`childCloser`) and full-update downwards (`GetNode`/`sync`)

### ISSUE THREAD

**Issue by schomatis** [ipfs/go-mfs#40]
Following the elucidation of the flush/sync mechanism in https://github.com/ipfs/go-mfs/issues/37 two different mechanism to update entries (when syncing/flushing a file) have been identified in MFS.

The first one is through the `childCloser` interface which provides an up-to-the-root update for a single file: when updating a file we update its parent directory, and this directory in turn updates *its* parent directory, and so on all the way up to the root. This is the kind of update specified by the user in the `ipfs files` command with the `--flush` option which would mean: after closing a file with modifications propagate them all the way up to the root so it's reflected in its CID (that is, at the UnixFS layer, the MFS at all moments has a reference to the up-to-date file in its internal state through `childDirs` and `files`).

Then there's a less documented general update mechanism when closing the MFS root (when the daemon shuts down or the offline program stops) which I'll term from now on as *full update*. The root's `Close` will call the `Directory`'s `GetNode()` method that calls the internal `sync()` method, calling in turn  `GetNode()` for all of its new/updated entries (and also `updateChild`, which is also part of the mechanism of the `closeChild` call chain),

https://github.com/ipfs/go-mfs/blob/2c1b835ceb3a181336651f2285dc4a13b9266c32/dir.go#L477-L491

https://github.com/ipfs/go-mfs/blob/2c1b835ceb3a181336651f2285dc4a13b9266c32/dir.go#L432-L439

So the interaction between `GetNode()` and `sync()` will trigger the same operations as the `closeChild` method uses to perform a up-to-the-root update for a single file but in this case it will be performed on all new/updated entries in the MFS hierarchy, effectively causing the full update.

This second method, full update, guarantees that if the root's `Close` is called (in normal circumstances it will, e.g., proper daemon shutdown or any offline command) no information will be lost and makes the first one, up-to-the-root update for a single file, useful only in particular cases (because of the performance penalty of doing updates any time a file is modified), e.g., when (with the MFS opened and under ongoing write/edit operations) we want to have the root's CID for publishing purposes (e.g., IPNS) pointing to the new version of the file that we're modifying (noting also that this update concerns only *that* file, any other modified file that wasn't "flushed" when it was closed won't be reflected in the root's CID until the final MFS full update).

This distinction should be clarified in the internal MFS code and also in the `go-ipfs` consumer commands like `ipfs files`.

**Comment by aschmahmann** [comment]
This repository is no longer maintained and has been copied over to [Boxo](https://github.com/ipfs/boxo).  In an effort to avoid noise and crippling in the Boxo repo from the weight of issues of the past, we are closing most issues and PRs in this repo.  Please feel free to open a new issue in Boxo (and reference this issue) if resolving this issue is still critical for unblocking or improving your usecase.

You can learn more in the [FAQs for the Boxo repo copying/consolidation effort](https://github.com/ipfs/boxo/wiki/Copied-or-Migrated-Repos-FAQ).
